### PR TITLE
Allows to manage image for devices 

### PR DESCRIPTION
**Pull request by cfrancisco**
* **Please check if the PR fulfills these requirements**
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
    This PR provides a new feature to GUI, allowing a user to handle images for devices. 
    In order to enable it, new interfaces have been added in edit area of templates, defining specific attributes on devices and allowing a CRUD for images. 
    Note that there're still some bugs, thus improvements will be developed as well as internationalization (i18)  and permissions.

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Is there any issue related to this PR in other repository?** (such as dojot/dojot)
This PR is connected to dojot/dojot#926.

* **Other information**:

### CODE REVIEW

File: src/js/actions/TemplateActions.js [hunk 1]
```diff
@@ -57,7 +79,7 @@ class TemplateActions {
 
     triggerUpdate(template, cb) {
         return (dispatch) => {
-            // console.log('triggerUpdate', template);
+            console.log('triggerUpdate', template);
```
Comment (mprevide) [inline src/js/actions/TemplateActions.js]: remove console log

File: src/js/actions/TemplateActions.js [hunk 2]
```diff
@@ -119,12 +145,16 @@ class TemplateActions {
         return error;
     }
 
-    selectTemplate(template = newTemplate){
+    selectTemplate(template = newTemplate) {
+        console.log("selectTemplate",template);
```
Comment (mprevide) [inline src/js/actions/TemplateActions.js]: remove console log

File: src/js/comms/templates/TemplateManager.js [hunk 3]
```diff
@@ -1,6 +1,8 @@
 /* eslint-disable */
```
Comment (mprevide) [inline src/js/comms/templates/TemplateManager.js]: Remove eslint-disable

File: src/js/actions/TemplateActions.js [hunk 4]
```diff
@@ -1,8 +1,9 @@
-/* eslint-disable */
+// /* eslint-disable */
```
Comment (mprevide) [inline src/js/actions/TemplateActions.js]: Remove eslint-disable

File: src/js/comms/templates/TemplateManager.js [hunk 5]
```diff
@@ -23,6 +25,21 @@ class TemplateManager {
...
     }
 
+    getTemplateGQL(id) {
+        const req = {
+            query: GQLTEMPLATE(id),
+        };
+        return util.POST(this.baseUrl+'/graphql/', req);
+    }
+
+    // getTemplateGQL(ID) {
```
Comment (mprevide) [inline src/js/comms/templates/TemplateManager.js]: remove comment code

File: src/js/comms/templates/TemplateManager.js [hunk 6]
```diff
@@ -45,3 +62,93 @@ class TemplateManager {
 
 const templateManager = new TemplateManager();
 export default templateManager;
+
+const GQLTEMPLATE2 = (templateId) =>`
```
Comment (mprevide) [inline src/js/comms/templates/TemplateManager.js]: Remove GQLTEMPLATE2

File: src/js/views/devices/Sidebar/SidebarImage/index.js [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+        if (!this.state.loaded && this.state.templateIdAllowedImage !== '')
+        {
+            const attrs = { current_state: 'Idle', update_result: 'No image', current_version: '0.1 Alpha' };
+            this.setState({
+                attrs,
+                loaded: true,
+            });
+
+
+            console.log('infos about devices', this.props.device);
```
Comment (mprevide) [inline src/js/views/devices/Sidebar/SidebarImage/index.js]: Remove

File: src/js/views/devices/Sidebar/index.js [hunk 8]
```diff
@@ -5,28 +5,49 @@ import AltContainer from 'alt-container';
...
+import ImageStore from 'Stores/ImageStore';
 import SidebarDevice from './SidebarDevice';
 import SidebarManageTemplates from './SidebarManageTemplates';
 import SidebarDeviceAttrs from './SidebarDeviceAttrs';
+import SidebarImage from './SidebarImage/index';
 import { FormActions } from '../Actions';
 
+
+function handleTemplateWithImages(device) {
+    console.log('Templates', device.templates);
```
Comment (mprevide) [inline src/js/views/devices/Sidebar/index.js]: Remove

File: src/js/views/devices/Sidebar/index.js [hunk 9]
```diff
@@ -309,25 +347,31 @@ class Sidebar extends Component {
...
             selectAttr,
             isNewDevice,
             errors,
             isShowSidebarDelete,
             deviceAttrsTitle,
+            templateIdAllowedImage,
+            hasTemplateWithImages,
         } = this.state;
+
+        console.log('hasTemplateWithImages', hasTemplateWithImages);
```
Comment (mprevide) [inline src/js/views/devices/Sidebar/index.js]: Remove

File: src/sass/_template.scss [hunk 10]
```diff
@@ -273,6 +279,8 @@
     //   color: rgba(0, 0, 0, 0.4);
```
Comment (mprevide) [inline src/sass/_template.scss]: Remove

File: src/sass/app.scss [hunk 11]
```diff
@@ -67,6 +67,12 @@ $details-background-color: #f3f3f3;
 // $main-font-family: Montserrat, sans-serif;
```
Comment (mprevide) [inline src/sass/app.scss]: remove

**Comment by mprevide** [comment]
- Remove commented console.logs and codes
- Place translation in pt and en (i18) in all texts
- Place permissions check in the required places
- Fix problems reported by codefactor and deepscan

**Comment by mprevide** [comment]
For add Permissions on firmware update:

- viewer just for GET
- modifier GET, POST, PUT and DELETE


import ability from 'Components/permissions/ability';
import Can from '../../components/permissions/Can';


const cannotEdit = !ability.can('modifier', 'fw-image');
const cannotEdit = !ability.can('viewer', 'fw-image');

or

<Can do="modifier" on="fw-image">
</Can>

<Can do="viewer" on="fw-image">
</Can>

**Comment by cfrancisco** [comment]
> For add Permissions on firmware update:
> 
> * viewer just for GET
> * modifier GET, POST, PUT and DELETE
> 
> import ability from 'Components/permissions/ability';
> import Can from '../../components/permissions/Can';
> 
> const cannotEdit = !ability.can('modifier', 'fw-image');
> const cannotEdit = !ability.can('viewer', 'fw-image');
> 
> or

Thanks! I'll add in the next PR.

**Comment by mprevide** [comment]
Put the translate in  locales/pt-br/templates.json

**Comment by mprevide** [comment]
The problems that codefactor points to seem simple, at least for the most part. Please fix. :)

**Comment by cfrancisco** [comment]
> The problems that codefactor points to seem simple, at least for the most part. Please fix. :)

Thanks for remembering, were fixed all possible points. Unfortunately, 16 issues are made by conceptual problems with Flux implementation and the propTypes.

**Comment by mprevide** [comment]
> Put the translate in locales/pt-br/templates.json

Don't forget too :)

**Comment by mprevide** [comment]
If you know there are still some bugs, why do not you fix them first? :)
I think it's better.

File: src/js/views/devices/Sidebar/SidebarImage/index.js [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+            const templateId = templateIdAllowedImage;
+            ImageActions.fetchImages.defer(templateId);
+            DeviceActions.fetchSingle.defer(deviceId, (device) => {
+                toaster.success('Device requested');
+                // We should ask data for 3 measures...
+                MeasureActions.fetchMeasure.defer(device, 'current_state', 1);
+                MeasureActions.fetchMeasure.defer(device, 'update_result', 1);
+                MeasureActions.fetchMeasure.defer(device, 'current_version', 1);
+            });
+            this.setState({
```
[the remark this answers is not in the document]
Comment (cfrancisco) [reply src/js/views/devices/Sidebar/SidebarImage/index.js]: Thanks for the advice.
For this point, please, also check the logic beyond the code snippet.
Suggestions for improvements in the code logic are welcome.

**Comment by cfrancisco** [comment]
> If you know there are still some bugs, why do not you fix them first? :)
> I think it's better.

Mainly because some bugs are also related to other Dojot components, such as device-manager or backstage as well as general GUI issues. Therefore, it is necessary to create issues for other components. 
At second, the bugs could be followed in separate issues, allowing us to track the expended time in each one, a better review and validation. 
Finally, the feature could to be tested by everyone, creating issues for any problems found.

**Comment by cfrancisco** [comment]
Duplicated PR with PR #384.

**Pull request closed without merging**